benchmarks/NN/functional_c#/Program.cs

```diff
@@ -0,0 +1,152 @@
+        static readonly ImmutableArray<ImmutableArray<double>> wheatData = rawData.Select(row => row.SetItem(row.Length - 1, row[^1] - 1)).ToImmutableArray();
+
+        static void Main(string[] args)
+        {
+            var nInput = wheatData[0].Length;
+            var nHidden = 5;
+            var nOutput = wheatData.Select(row => row[row.Length - 1]).Distinct().Count();
+            var learningRate = 0.3;
+            var epochs = 500;
+
+            var dataset = normaliseDataset(wheatData, datasetMinMax(wheatData))
+                        .OrderBy(_ => rand.Next())
+                        .ToImmutableArray();
+            
+            var splitIndex = dataset.Length / 10;
+            var testData = dataset.Take(splitIndex).ToImmutableArray();
+            var trainData = dataset.TakeLast(dataset.Length - splitIndex).ToImmutableArray();
+
+            var init = initialiseNetwork(nInput, nHidden, nOutput);
+
+            var trainedNetwork = trainNetwork(init, trainData, learningRate, epochs, nOutput);
+ 
+            var results = testData.Select(row => (prediction: predict(trainedNetwork, row), actual: (int)row[^1])).ToImmutableArray();
+            var correctPredictions = results.Where(res => res.prediction == res.actual).Count();
+            var accuracy = correctPredictions / (double)results.Length * 100.0;
+            
+            System.Console.WriteLine(accuracy);
+        }
+
+        private static ImmutableArray<Layer> backwardsPropagateError(ImmutableArray<Layer> network, ImmutableArray<double> expected)
+        {
+            var firstLayer = network.First();
+            var remainingLayers = network.TakeLast(network.Length - 1);
+            var errors = firstLayer.Zip(expected, (neuron, exp) => exp - neuron.output).ToImmutableArray();
+            
+            Func<double, double> derivative = (neuronOutput) => neuronOutput * (1.0 - neuronOutput);
+            Func<double, Neuron, Neuron> getError = (error, neuron) => new Neuron(neuron.weights, neuron.output, neuron.bias, error * derivative(neuron.output));
+            Func<ImmutableArray<double>, Layer, Layer> delta = (errors, layer) => errors.Zip(layer, (error, neuron) => getError(error, neuron)).ToImmutableArray();
+            Func<Layer, int, double> neuronError = (layer, i) => layer.Select(neuron => neuron.weights[i] * neuron.delta).Sum();
+            Func<ImmutableArray<Layer>, Layer, ImmutableArray<Layer>> backProp = (res, layer) => {
+                var errors = layer.Select((neuron, i) => neuronError(res[0], i)).ToImmutableArray();
+                var newLayer = delta(errors, layer);
+                return res.Add(newLayer);
+            };
+
+            return remainingLayers.Aggregate(ImmutableArray<Layer>.Empty.Add(delta(errors, firstLayer)), (acc, r) => backProp(acc, r).Reverse().ToImmutableArray());

```

Comment:
Overvej om det er nødvendigt at have den her reverse, og om der er for meget med i fold funktionen.